### PR TITLE
PHOENIX-7229 Leverage bloom filters for single key point lookups

### PR DESCRIPTION
**Pull request by tkhurana**
Jira: PHOENIX-7229

### CODE REVIEW

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/BloomFilterIT.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
+package org.apache.phoenix.end2end;
```
Comment (virajjasani) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/BloomFilterIT.java:1]: ASF lincense

**Comment by virajjasani** [comment]
@tkhurana the reason why this PR has build failure is because of `MetricsRegionServerWrapper` usage. It comes from `hbase-hadoop-compat`, otherwise the master/5.2 branches don't have that issue.

**Review by kadirozde — APPROVED** [review]
+1, Thanks

**Comment by tkhurana** [comment]
> Need updates in WhereOptimizerTest, since we are updating high level Scan object's stop rowkey. Updated in the latest revision.

@virajjasani No, actually scanFromContext is not changing. It is the scan we generate from the iterator is changing.

**Comment by virajjasani** [comment]
I made this commit https://github.com/apache/phoenix/commit/1e3fa9dae41c361ba2b7f4b381ecbf08568ee126
For single pointlookup, scan context can also be changed. Does it hurt?

**Comment by virajjasani** [comment]
Earlier also scan context object had same scan boundary as the scans generated from it.

File: phoenix-core-client/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java [hunk 2]
```diff
@@ -941,12 +940,7 @@ private ScansWithRegionLocations getParallelScans(byte[] startKey, byte[] stopKe
             Scan scanFromContext = context.getScan();
             if (scanRanges.getPointLookupCount() == 1) {
                 // leverage bloom filter for single key point lookup by turning scan to Get Scan#isGetScan()
-                try {
-                    scanFromContext = new Scan(context.getScan());
-                } catch (IOException e) {
-                    throw new PhoenixIOException(e);
-                }
-                scanFromContext.withStopRow(scanFromContext.getStartRow(), scanFromContext.includeStartRow());
+                scanFromContext.withStopRow(scanFromContext.getStartRow(), true);
```
Comment (tkhurana) [inline phoenix-core-client/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:943]: I actually didn't want to change the scan stored in StatementContext. I wanted to modify the scan we generate in getIterators.

Comment (tkhurana) [reply phoenix-core-client/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:943]: You generate a scan in StatementContext and then we generate the actual scans when we are creating the iterators like a high level scan in statementcontext could generate multiple parallel scans one for each region server. That is why I was copying the original scan stored in StatementContext.

Comment (virajjasani) [reply phoenix-core-client/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:943]: I see, then we can revert it.

Comment (tkhurana) [reply phoenix-core-client/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java:943]: Earlier the scan object had stop key as next(start key) for point lookups. Take a look at this test case https://github.com/apache/phoenix/blob/master/phoenix-core/src/test/java/org/apache/phoenix/compile/WhereOptimizerTest.java#L988-L992

**Comment by tkhurana** [comment]
> @tkhurana the reason why this PR has build failure is because of `MetricsRegionServerWrapper` usage. It comes from `hbase-hadoop-compat`, otherwise the master/5.2 branches don't have that issue.

Thanks @virajjasani It makes sense.

**Comment by virajjasani** [comment]
@tkhurana HBASE-27241 is only present on hbase 2.4 based build fails if we try to access bloom filter metrics with IT test, i will have to revert the IT test, we can figure out how best to write test later.

**Pull request merged**